### ISSUE TITLE
Compile error with 7.10.2 against 0.7.9.0.

### ISSUE THREAD

**Issue by seanparsons** [agrafix/Spock#38]
I'm not sure if this is a 7.10.2 issue or a dependency bounds issue, the same project works built against 7.8.4:

```
Downloading ansi-terminal-0.6.2.1...
Downloading appar-0.1.4...
Configuring base-orphans-0.3.2...
Configuring base64-bytestring-1.0.0.1...
Configuring byteable-0.1.1...
Configuring data-default-class-0.0.1...
Configuring dlist-0.7.1.1...
Configuring mmorph-1.0.4...
Configuring nats-1...
Configuring mtl-2.2.1...
Downloading auto-update-0.1.2.1...
Building mmorph-1.0.4...
Building base-orphans-0.3.2...
Building data-default-class-0.0.1...
Building nats-1...
Building dlist-0.7.1.1...
Building byteable-0.1.1...
Building mtl-2.2.1...
Building base64-bytestring-1.0.0.1...
Downloading base16-bytestring-0.1.1.6...
Downloading byteorder-1.0.4...
Configuring network-2.6.1.0...
Installed nats-1
Installed data-default-class-0.0.1
Configuring old-locale-1.0.0.7...
Configuring prelude-extras-0.4...
Installed base-orphans-0.3.2
Configuring random-1.1...
Installed byteable-0.1.1
Building old-locale-1.0.0.7...
Building prelude-extras-0.4...
Configuring primitive-0.6...
Installed mmorph-1.0.4
Building random-1.1...
Downloading bytestring-builder-0.10.6.0.0...
Configuring stm-2.4.4...
Installed dlist-0.7.1.1
Downloading easy-file-0.2.1...
Configuring syb-0.4.4...
Installed base64-bytestring-1.0.0.1
Downloading network-info-0.2.0.6...
Building primitive-0.6...
Downloading parallel-3.2.0.6...
Downloading reflection-1.5.2.1...
Building stm-2.4.4...
Downloading safe-0.3.9...
Downloading stringsearch-0.3.6.6...
Building syb-0.4.4...
Downloading unix-compat-0.4.1.4...
Configuring tagged-0.8.0.1...
Installed old-locale-1.0.0.7
Downloading word8-0.1.2...
Configuring text-1.2.1.0...
Downloading data-default-instances-base-0.0.1...
Installed mtl-2.2.1
Downloading data-default-instances-containers-0.0.1...
Building tagged-0.8.0.1...
Configuring transformers-compat-0.4.0.4...
Installed prelude-extras-0.4
Downloading data-default-instances-dlist-0.0.1...
Downloading data-default-instances-old-locale-0.0.1...
Building text-1.2.1.0...
Downloading regex-base-0.93.2...
Downloading wiring-0.3.0...
Building transformers-compat-0.4.0.4...
Configuring zlib-0.6.1.1...
Installed stm-2.4.4
Building network-2.6.1.0...
Configuring appar-0.1.4...
Installed transformers-compat-0.4.0.4
Building zlib-0.6.1.1...
Building appar-0.1.4...
Configuring auto-update-0.1.2.1...
Installed primitive-0.6
Configuring ansi-terminal-0.6.2.1...
Installed tagged-0.8.0.1
Installed syb-0.4.4
Configuring cryptohash-0.11.6...
Building auto-update-0.1.2.1...
Building ansi-terminal-0.6.2.1...
Configuring base16-bytestring-0.1.1.6...
Installed appar-0.1.4
Building cryptohash-0.11.6...
Configuring byteorder-1.0.4...
Installed random-1.1
Building base16-bytestring-0.1.1.6...
Configuring bytestring-builder-0.10.6.0.0...
Installed auto-update-0.1.2.1
Building byteorder-1.0.4...
Building bytestring-builder-0.10.6.0.0...
Configuring easy-file-0.2.1...
Installed base16-bytestring-0.1.1.6
Configuring network-info-0.2.0.6...
Installed bytestring-builder-0.10.6.0.0
Configuring reflection-1.5.2.1...
Installed byteorder-1.0.4
Building easy-file-0.2.1...
Building network-info-0.2.0.6...
Building reflection-1.5.2.1...
Configuring parallel-3.2.0.6...
Installed easy-file-0.2.1
Configuring safe-0.3.9...
Installed ansi-terminal-0.6.2.1
Configuring old-time-1.1.0.3...
Installed reflection-1.5.2.1
Building parallel-3.2.0.6...
Configuring stringsearch-0.3.6.6...
Installed network-info-0.2.0.6
Building safe-0.3.9...
Building stringsearch-0.3.6.6...
Configuring word8-0.1.2...
Installed zlib-0.6.1.1
Configuring data-default-instances-base-0.0.1...
Installed parallel-3.2.0.6
Building word8-0.1.2...
Configuring data-default-instances-containers-0.0.1...
Installed safe-0.3.9
Building old-time-1.1.0.3...
Building data-default-instances-base-0.0.1...
Building data-default-instances-containers-0.0.1...
Configuring data-default-instances-old-locale-0.0.1...
Installed word8-0.1.2
Configuring data-default-instances-dlist-0.0.1...
Installed network-2.6.1.0
Downloading iproute-1.4.0...
Downloading simple-sendfile-0.2.20...
Configuring unix-compat-0.4.1.4...
Installed data-default-instances-base-0.0.1
Building data-default-instances-old-locale-0.0.1...
Configuring regex-base-0.93.2...
Installed data-default-instances-containers-0.0.1
Building data-default-instances-dlist-0.0.1...
Building unix-compat-0.4.1.4...
Configuring wiring-0.3.0...
Installed data-default-instances-old-locale-0.0.1
Building regex-base-0.93.2...
Configuring StateVar-1.1.0.0...
Installed cryptohash-0.11.6
Configuring transformers-base-0.4.4...
Installed data-default-instances-dlist-0.0.1
Downloading data-default-0.5.3...
Building wiring-0.3.0...
Building StateVar-1.1.0.0...
Building transformers-base-0.4.4...
Configuring exceptions-0.8.0.2...
Installed unix-compat-0.4.1.4
Configuring vector-0.10.12.3...
Installed StateVar-1.1.0.0
Building exceptions-0.8.0.2...
Configuring distributive-0.4.4...
Installed transformers-base-0.4.4
Installed regex-base-0.93.2
Downloading regex-posix-0.95.2...
Configuring tf-random-0.5...
Building vector-0.10.12.3...
Building tf-random-0.5...
Configuring simple-sendfile-0.2.20...
Installed old-time-1.1.0.3
Downloading unix-time-0.3.5...
Building simple-sendfile-0.2.20...
Configuring data-default-0.5.3...
Installed wiring-0.3.0
Configuring monad-control-1.0.0.4...
Installed exceptions-0.8.0.2
Building data-default-0.5.3...
Building monad-control-1.0.0.4...
Configuring iproute-1.4.0...
Installed simple-sendfile-0.2.20
Configuring regex-posix-0.95.2...
Installed stringsearch-0.3.6.6
Configuring unix-time-0.3.5...
Installed data-default-0.5.3
Building iproute-1.4.0...
Building regex-posix-0.95.2...
Installed tf-random-0.5
Configuring QuickCheck-2.8.1...
Installed monad-control-1.0.0.4
Configuring lifted-base-0.2.3.6...
Building QuickCheck-2.8.1...
Building lifted-base-0.2.3.6...
Building distributive-0.4.4...
Building unix-time-0.3.5...
Installed distributive-0.4.4
Installed lifted-base-0.2.3.6
Configuring resourcet-1.1.5...
Building resourcet-1.1.5...
Installed unix-time-0.3.5
Installed iproute-1.4.0
Installed regex-posix-0.95.2
Downloading regex-compat-0.95.1...
Configuring regex-compat-0.95.1...
Building regex-compat-0.95.1...
Installed resourcet-1.1.5
Configuring conduit-1.2.4.2...
Building conduit-1.2.4.2...
Installed regex-compat-0.95.1
Installed conduit-1.2.4.2
Installed QuickCheck-2.8.1
Installed text-1.2.1.0
Configuring blaze-builder-0.4.0.1...
Downloading clay-0.10.1...
Downloading digestive-functors-0.8.0.0...
Configuring hashable-1.2.3.2...
Configuring mime-types-0.1.0.6...
Configuring parsec-3.1.9...
Building hashable-1.2.3.2...
Building parsec-3.1.9...
Building mime-types-0.1.0.6...
Building blaze-builder-0.4.0.1...
Downloading fast-logger-2.3.1...
Configuring clay-0.10.1...
Configuring digestive-functors-0.8.0.0...
Downloading path-pieces-0.2.0...
Configuring path-pieces-0.2.0...
Downloading xml-types-0.3.4...
Building clay-0.10.1...
Building digestive-functors-0.8.0.0...
Building path-pieces-0.2.0...
Installed hashable-1.2.3.2
Configuring fast-logger-2.3.1...
Downloading uuid-types-1.0.1...
Building fast-logger-2.3.1...
Configuring xml-types-0.3.4...
Installed blaze-builder-0.4.0.1
Downloading blaze-markup-0.7.0.2...
Installed path-pieces-0.2.0
Configuring case-insensitive-1.2.0.4...
Building xml-types-0.3.4...
Building case-insensitive-1.2.0.4...
Configuring scientific-0.3.3.8...
Installed fast-logger-2.3.1
Configuring uuid-types-1.0.1...
Installed digestive-functors-0.8.0.0
Building scientific-0.3.3.8...
Configuring unordered-containers-0.2.5.1...
Installed case-insensitive-1.2.0.4
Building uuid-types-1.0.1...
Configuring cookie-0.4.1.5...
Installed vector-0.10.12.3
Downloading resource-pool-0.2.3.2...
Building unordered-containers-0.2.5.1...
Building cookie-0.4.1.5...
Configuring streaming-commons-0.1.12.1...
Installed mime-types-0.1.0.6
Installed cookie-0.4.1.5
Configuring blaze-markup-0.7.0.2...
Building streaming-commons-0.1.12.1...
Configuring http-types-0.8.6...
Installed parsec-3.1.9
Building blaze-markup-0.7.0.2...
Configuring resource-pool-0.2.3.2...
Installed uuid-types-1.0.1
Downloading uuid-1.3.10...
Building http-types-0.8.6...
Building resource-pool-0.2.3.2...
Installed scientific-0.3.3.8
Configuring network-uri-2.6.0.3...
Configuring uuid-1.3.10...
Installed xml-types-0.3.4
Building network-uri-2.6.0.3...
Building uuid-1.3.10...
Configuring attoparsec-0.13.0.0...
Installed resource-pool-0.2.3.2
Building attoparsec-0.13.0.0...
Installed http-types-0.8.6
Installed uuid-1.3.10
Installed clay-0.10.1
Installed blaze-markup-0.7.0.2
Downloading blaze-html-0.8.0.2...
Installed unordered-containers-0.2.5.1
Downloading graph-core-0.2.2.0...
Configuring semigroups-0.16.2.2...
Installed streaming-commons-0.1.12.1
Downloading vault-0.3.0.4...
Configuring blaze-html-0.8.0.2...
Configuring graph-core-0.2.2.0...
Building semigroups-0.16.2.2...
Configuring vault-0.3.0.4...
Building blaze-html-0.8.0.2...
Building graph-core-0.2.2.0...
Building vault-0.3.0.4...
Installed network-uri-2.6.0.3
Configuring http-client-0.4.11.2...
Building http-client-0.4.11.2...
Installed vault-0.3.0.4
Downloading wai-3.0.2.3...
Installed graph-core-0.2.2.0
Downloading reroute-0.2.3.0...
Configuring wai-3.0.2.3...
Configuring reroute-0.2.3.0...
Building wai-3.0.2.3...
Building reroute-0.2.3.0...
Installed wai-3.0.2.3
Downloading wai-logger-2.2.4...
Configuring wai-logger-2.2.4...
Building wai-logger-2.2.4...
Installed semigroups-0.16.2.2
Configuring bifunctors-5...
Configuring void-0.7...
Building void-0.7...
Building bifunctors-5...
Installed wai-logger-2.2.4
Installed void-0.7
Configuring contravariant-1.3.1.1...
Downloading wai-extra-3.0.7.1...
Configuring wai-extra-3.0.7.1...
Building contravariant-1.3.1.1...
Installed reroute-0.2.3.0
Building wai-extra-3.0.7.1...
Installed contravariant-1.3.1.1
Configuring comonad-4.2.6...
Installed http-client-0.4.11.2
Installed bifunctors-5
Installed blaze-html-0.8.0.2
Downloading digestive-functors-blaze-0.6.0.6...
Building comonad-4.2.6...
Configuring digestive-functors-blaze-0.6.0.6...
Building digestive-functors-blaze-0.6.0.6...
Installed attoparsec-0.13.0.0
Downloading aeson-0.9.0.1...
Downloading conduit-extra-1.1.8...
Installed wai-extra-3.0.7.1
Installed digestive-functors-blaze-0.6.0.6
Downloading http-date-0.0.6.1...
Configuring conduit-extra-1.1.8...
Installed comonad-4.2.6
Configuring profunctors-5.1.1...
Configuring semigroupoids-5.0.0.1...
Building conduit-extra-1.1.8...
Building profunctors-5.1.1...
Configuring http-date-0.0.6.1...
Configuring aeson-0.9.0.1...
Building http-date-0.0.6.1...
Building aeson-0.9.0.1...
Building semigroupoids-5.0.0.1...
Installed http-date-0.0.6.1
Downloading warp-3.0.13.1...
Configuring warp-3.0.13.1...
Installed profunctors-5.1.1
Building warp-3.0.13.1...
Installed conduit-extra-1.1.8
Downloading xml-conduit-1.3.0...
Configuring xml-conduit-1.3.0...
Building xml-conduit-1.3.0...
Installed semigroupoids-5.0.0.1
Configuring free-4.12.1...
Building free-4.12.1...
Installed warp-3.0.13.1
Installed xml-conduit-1.3.0
Installed free-4.12.1
Downloading adjunctions-4.2.1...
Configuring adjunctions-4.2.1...
Building adjunctions-4.2.1...
Installed adjunctions-4.2.1
Downloading kan-extensions-4.2.2...
Configuring kan-extensions-4.2.2...
Installed aeson-0.9.0.1
Downloading Spock-0.7.9.0...
Building kan-extensions-4.2.2...
Configuring Spock-0.7.9.0...
Building Spock-0.7.9.0...
Installed kan-extensions-4.2.2
Downloading lens-4.11...
Failed to install Spock-0.7.9.0
Build log ( /home/sean/workspace/oneaday/.cabal-sandbox/logs/Spock-0.7.9.0.log ):
Configuring Spock-0.7.9.0...
Building Spock-0.7.9.0...
Preprocessing library Spock-0.7.9.0...
[ 1 of 10] Compiling Web.Spock.Internal.Wire ( src/Web/Spock/Internal/Wire.hs, dist/dist-sandbox-3e6995ac/build/Web/Spock/Internal/Wire.o )
[ 2 of 10] Compiling Web.Spock.Internal.Util ( src/Web/Spock/Internal/Util.hs, dist/dist-sandbox-3e6995ac/build/Web/Spock/Internal/Util.o )
[ 3 of 10] Compiling Web.Spock.Internal.CoreAction ( src/Web/Spock/Internal/CoreAction.hs, dist/dist-sandbox-3e6995ac/build/Web/Spock/Internal/CoreAction.o )
[ 4 of 10] Compiling Web.Spock.Internal.Types ( src/Web/Spock/Internal/Types.hs, dist/dist-sandbox-3e6995ac/build/Web/Spock/Internal/Types.o )

src/Web/Spock/Internal/Types.hs:143:80:
    Couldn't match expected type ‘ReaderT
                                    (WebState conn sess st) (ResourceT IO) a1
                                  -> IO a1’
                with actual type ‘forall a2.
                                  ReaderT (WebState conn sess st) (ResourceT IO) a2
                                  -> IO (StM (ReaderT (WebState conn sess st) (ResourceT IO)) a2)’
    Relevant bindings include
      runInBase :: forall a.
                   ReaderT (WebState conn sess st) (ResourceT IO) a
                   -> IO (StM (ReaderT (WebState conn sess st) (ResourceT IO)) a)
        (bound at src/Web/Spock/Internal/Types.hs:143:50)
      f :: RunInBase (WebStateM conn sess st) IO -> IO a
        (bound at src/Web/Spock/Internal/Types.hs:143:18)
      liftBaseWith :: (RunInBase (WebStateM conn sess st) IO -> IO a)
                      -> WebStateM conn sess st a
        (bound at src/Web/Spock/Internal/Types.hs:143:5)
    In the first argument of ‘(.)’, namely ‘runInBase’
    In the second argument of ‘(.)’, namely ‘runInBase . runWebStateM’
```

**Comment by agrafix** [comment]
This looks like an issue with monad-control and/or ghc 7.10.2. We are not using ghc 7.10 yet, but I'll see if I can reproduce the issue. It's still interesting because the travis build for ghc7.10 and the stackage build seem to work fine...

**Comment by jml** [comment]
I ran into this with 7.10.2 but *not* 7.10.1. See https://travis-ci.org/jml/hazard/jobs/65775138.

**Comment by agrafix** [comment]
Yes, I can confirm that it only happens when building 7.10.2. Not sure why yet...

**Comment by agrafix** [comment]
Does this fail with RC1, too? Can you test that? @jml / @seanparsons

**Comment by jml** [comment]
Unlikely to test any time soon.

**Comment by dnhgff** [comment]
i have reproduced this with 7.10.2 RC1

**Comment by elfeck** [comment]
I also reproduced this with
```
ghc --version
The Glorious Glasgow Haskell Compilation System, version 7.10.1.20150614
```

**Comment by agrafix** [comment]
Expect a fix this week! Still not sure why travis doesn't find the issue :-1:

**Comment by agrafix** [comment]
Should be fixed since ffd46ce1c8cb0b9e3edf7b652fb3a3147bb10eaa and the new 0.7.10 version.

**Comment by agrafix** [comment]
If the problem still exists, please reopen
